src/transform.ts

```diff
@@ -47,7 +46,13 @@ export function addCoersion(def: any): void {
   }
 }
 
+const sanitizeTemplate = (s: string) => s.replace(/<([^>]+)>/, (_, m) => `_of_${m.replace(/,\s*/g, '_')}`);
+const parensTypeToString = (def: TypeDef) => `(${typeToString(def)})`;
```

Comment:
This transformation needs to recurse down the template vars, but even then it will very-many-to-one.  Consider e.g. (with non-trivial recursion, it needs at least to match up `<`...`>` pairs)
```ts
type TwoOrThree<A, B, C=number> = [A, B, C];
type Foo = TwoOrThree<number, TwoOrThree<null, string, number>>;  // TwoOrThree_of_number_TwoOrThree_of_null_string_number
type Bar = TwoOrThree<number, TwoOrThree<null, string>, number>;  // TwoOrThree_of_number_TwoOrThree_of_null_string_number
```
We can parse using recursive descent (or any other way; I think you can parse a great many types using just lookahead 1 over individual characters).  But we'll still need a really good way of constructing the type name on the other side.
A very simple alternative would be to encode ASCII as string names; that also makes it a bit easier for humans.  E.g.

| from | to |
|---|---|
| `<` | `_of_` |
| `>` | `_end` |
| `,` | `_` |
| `[` | `_array_of_` |
| ']' | `_end` |

using the (unencoded!) fact that `<`...`>` and `[`...`]` nest.  This still has the problem that `of`, `array` and `end` are legal type names, but I reckon we can just add a configuration option if it ever goes that far.